src/pokedex_data.c

```diff
@@ -0,0 +1,1235 @@
+#include "pokedex_data.h"
+
+#include <nitro.h>
+#include <string.h>
+
+#include "constants/species.h"
+
+#include "heap.h"
+#include "inlines.h"
+#include "pokemon.h"
+#include "savedata.h"
+#include "unk_020986CC.h"
+
+#define DEX_SIZE_U32           ((int)((NATIONAL_DEX_COUNT - 1) / 32) + 1) // default 16
+#define MAGIC_NUMBER           0xbeefcafe
+#define NUM_MYTHICALS_NATIONAL 11
+#define NATIONAL_DEX_GOAL      (NATIONAL_DEX_COUNT - NUM_MYTHICALS_NATIONAL)
+#define UNOWN_COUNT            28
+#define DEOXYS_COUNT           4
+#define ROTOM_COUNT            6
+#define TERMINAL_4BITS         0x7
+#define TERMINAL_BYTE          0xf
+#define TERMINAL_U8            0xff
+#define TERMINAL_U32           0xffffffff
+
+typedef struct PokedexData {
+    u32 magic;
+    u32 caughtPokemon[DEX_SIZE_U32];
+    u32 seenPokemon[DEX_SIZE_U32];
+    u32 recordedGenders[2][DEX_SIZE_U32];
+    u32 spindaForm;
+    u8 shellosFormsSeen;
+    u8 gastrodonFormsSeen;
+    u8 burmyFormsSeen;
+    u8 wormadamFormsSeen;
+    u8 unownFormsSeen[UNOWN_COUNT];
+    u8 recordedLanguages[MAX_SPECIES + 1];
+    u8 canDetectForms;
+    u8 canDetectLanguages;
+    u8 pokedexObtained;
+    u8 nationalDexObtained;
+    u32 rotomFormsSeen;
+    u8 shayminFormsSeen;
+    u8 giratinaFormsSeen;
+} PokedexData;
+
+int PokedexData_SaveSize(void)
+{
+    return sizeof(PokedexData);
+}
+
+PokedexData *PokedexData_Alloc(u32 heapID)
+{
+    PokedexData *pokedexData = Heap_AllocFromHeap(heapID, sizeof(PokedexData));
+    PokedexData_Init(pokedexData);
+
+    return pokedexData;
+}
+
+void PokedexData_Copy(const PokedexData *src, PokedexData *dest)
+{
+    MI_CpuCopy8(src, dest, sizeof(PokedexData));
+}
+
+static inline void CheckPokedexIntegrity(const PokedexData *pokedexData)
+{
+    GF_ASSERT(pokedexData->magic == MAGIC_NUMBER);
+}
+
+static BOOL SpeciesInvalid(u16 species)
+{
+    if (species == SPECIES_NONE || species > NATIONAL_DEX_COUNT) {
+        GF_ASSERT(FALSE);
+        return TRUE;
+    } else {
+        return FALSE;
+    }
+}
+
+static inline BOOL ReadBit_2Forms(const u8 *array, u16 bitIndex)
+{
+    bitIndex--;
+    return 0 != (array[bitIndex >> 0x3] & (1 << (bitIndex & 0x7)));
```

Comment:
**style**: Use decimal values for bitshift operands; applies throughout the file.

```suggestion
    return 0 != (array[bitIndex >> 3] & (1 << (bitIndex & 0x7)));
```